### ISSUE TITLE
Slicing may be incorrect

### ISSUE THREAD

**Issue by zshaheen** [E3SM-Project/e3sm_diags#206]
Ex: When slicing from 1990 to 1992, running `test.getTime().asComponentTime()` shows the following:
```
[1990-2-1 0:0:0.0, 1990-3-1 0:0:0.0, 1990-4-1 0:0:0.0, 1990-5-1 0:0:0.0,
1990-6-1 0:0:0.0, 1990-7-1 0:0:0.0, 1990-8-1 0:0:0.0, 1990-9-1 0:0:0.0,
1990-10-1 0:0:0.0, 1990-11-1 0:0:0.0, 1990-12-1 0:0:0.0, 1991-1-1 0:0:0.0,
1991-2-1 0:0:0.0, 1991-3-1 0:0:0.0, 1991-4-1 0:0:0.0, 1991-5-1 0:0:0.0,
1991-6-1 0:0:0.0, 1991-7-1 0:0:0.0, 1991-8-1 0:0:0.0, 1991-9-1 0:0:0.0,
1991-10-1 0:0:0.0, 1991-11-1 0:0:0.0, 1991-12-1 0:0:0.0, 1992-1-1 0:0:0.0,
1992-2-1 0:0:0.0, 1992-3-1 0:0:0.0, 1992-4-1 0:0:0.0, 1992-5-1 0:0:0.0,
1992-6-1 0:0:0.0, 1992-7-1 0:0:0.0, 1992-8-1 0:0:0.0, 1992-9-1 0:0:0.0,
1992-10-1 0:0:0.0, 1992-11-1 0:0:0.0, 1992-12-1 0:0:0.0, 1993-1-1 0:0:0.0]
```

It seems like we're off by one time slice. This only happens with the test data and not the ref data. Thanks you @chengzhuzhang for finding this.

**Comment by zshaheen** [comment]
A better way to slice is to do something like the below:
```python
path = 'PRECT_197901_198312.nc'
with cdms2.open(path) as f:
    start = '1960'
    end = '1960'
    s_time = '{}-1'.format(start)
    e_time = '{}-12'.format(end)
    var = f('PRECT', time=(s_time, e_time))
```

The way we calculate the slice was error prone. The good news is that if we pass in an incorrect slice, it catches that error.
```
cdms2.error.CDMSError: Coordinate interval is out of range or intersection has no data: ('1960-1', '1960-12')
```

**Comment by zshaheen** [comment]
The above new implementation was done in 71d9a2c995acbd875fa2d32ea9e44cc02893114b. However, we are still getting some incorrect results.

I'm running [this example with time series data on NERSC](https://e3sm-project.github.io/e3sm_diags/docs/html/examples/model_ts-vs-model_ts.html). We see that we are just comparing different time slices from the same data, 1850-1852 vs 2011-2013. When doing so, here's what the time slices are:
```
[2011-1-1 0:0:0.0, 2011-2-1 0:0:0.0, 2011-3-1 0:0:0.0, 2011-4-1 0:0:0.0, 2011-5-1 0:0:0.0,
2011-6-1 0:0:0.0, 2011-7-1 0:0:0.0, 2011-8-1 0:0:0.0, 2011-9-1 0:0:0.0, 2011-10-1 0:0:0.0,
2011-11-1 0:0:0.0, 2011-12-1 0:0:0.0, 2012-1-1 0:0:0.0, 2012-2-1 0:0:0.0, 2012-3-1 0:0:0.0,
2012-4-1 0:0:0.0, 2012-5-1 0:0:0.0, 2012-6-1 0:0:0.0, 2012-7-1 0:0:0.0, 2012-8-1 0:0:0.0,
2012-9-1 0:0:0.0, 2012-10-1 0:0:0.0, 2012-11-1 0:0:0.0, 2012-12-1 0:0:0.0, 2013-1-1 0:0:0.0,
2013-2-1 0:0:0.0, 2013-3-1 0:0:0.0, 2013-4-1 0:0:0.0, 2013-5-1 0:0:0.0, 2013-6-1 0:0:0.0,
2013-7-1 0:0:0.0, 2013-8-1 0:0:0.0, 2013-9-1 0:0:0.0, 2013-10-1 0:0:0.0, 2013-11-1 0:0:0.0,
2013-12-1 0:0:0.0]

[1850-2-1 0:0:0.0, 1850-3-1 0:0:0.0, 1850-4-1 0:0:0.0, 1850-5-1 0:0:0.0, 1850-6-1 0:0:0.0,
1850-7-1 0:0:0.0, 1850-8-1 0:0:0.0, 1850-9-1 0:0:0.0, 1850-10-1 0:0:0.0, 1850-11-1 0:0:0.0,
1850-12-1 0:0:0.0, 1851-1-1 0:0:0.0, 1851-2-1 0:0:0.0, 1851-3-1 0:0:0.0, 1851-4-1 0:0:0.0,
1851-5-1 0:0:0.0, 1851-6-1 0:0:0.0, 1851-7-1 0:0:0.0, 1851-8-1 0:0:0.0, 1851-9-1 0:0:0.0,
1851-10-1 0:0:0.0, 1851-11-1 0:0:0.0, 1851-12-1 0:0:0.0, 1852-1-1 0:0:0.0, 1852-2-1 0:0:0.0,
1852-3-1 0:0:0.0, 1852-4-1 0:0:0.0, 1852-5-1 0:0:0.0, 1852-6-1 0:0:0.0, 1852-7-1 0:0:0.0,
1852-8-1 0:0:0.0, 1852-9-1 0:0:0.0, 1852-10-1 0:0:0.0, 1852-11-1 0:0:0.0, 1852-12-1 0:0:0.0]
```

We see that **when slicing data from the start of the file, the first month is missing**. So we don't see 1850-1-1.

@chengzhuzhang, all of the data is located on NERSC here:
```
/global/project/projectdirs/acme/acme_diags/test_model_data_for_acme_diags/time-series/E3SM_v1/
```
Is something wrong with the data? A quick examination doesn't make it appear so, since the units for the `time` are `'days since 1850-01-01 00:00:00'`.

**Comment by doutriaux1** [comment]
@zshaheen I would do:
```python
path = 'PRECT_197901_198312.nc'
with cdms2.open(path) as f:
    start = '1960'
    end = '1961'
    var = f('PRECT', time=(start,end,'con'))
```

**Comment by doutriaux1** [comment]
@zshaheen ONE caveat on the above one is if (and I believe it's the case for E3SM since it comes from NCAR) the time is stored at the END of the month. In which case everything "looks" one month off.
you might need to run
```python
cdutil.setTimeBoundsMonthly(data,stored=1)
```

**Comment by zshaheen** [comment]
@doutriaux1, I've tried what you suggested and I'm still getting the same issue, the data when sliced from the beginning of the file is off by one month. And it's just not just the labeling, the data itself is missing a slice.

**Comment by doutriaux1** [comment]
@zshaheen ping me when you get in, I'll take a look with you

**Comment by zshaheen** [comment]
Here is code to create the issue. I'm using cdms 3.1.2, which again shows as version 3.0.0 under Anaconda.

The data is located on acme1/aims4. There's a path for the data on NERSC, by Cori is down and Edison has passed away.

```python
# https://github.com/E3SM-Project/e3sm_diags/issues/206

import os
import cdms2

path = '/p/cscratch/acme/data/test_model_data_for_acme_diags/time-series/E3SM_v1'
# For NERSC, use the path below:
# path = '/global/project/projectdirs/acme/acme_diags/test_model_data_for_acme_diags/time-series/E3SM_v1/'
file_name = 'PRECC_185001_201312.nc'
var = 'PRECC'

with cdms2.open(os.path.join(path, file_name)) as f:
    start, end = 1850, 1850
    s_time = '{}-1'.format(start)
    e_time = '{}-12'.format(end)
    v = f(var, time=(s_time, e_time))
    times = v.getTime().asComponentTime()
    print('The num of times were {} when it should be {}.'.format(len(times), (end-start+1)*12))
    print(times)
```

**Comment by zshaheen** [comment]
In the above code, when we use `v = f(var, time=slice(0, 12))`, I'm still not getting January of 1850, but am getting January 1851, since I'm getting 12 slices. See below:
```
[1850-2-1 0:0:0.0, 1850-3-1 0:0:0.0, 1850-4-1 0:0:0.0, 1850-5-1 0:0:0.0,
1850-6-1 0:0:0.0, 1850-7-1 0:0:0.0, 1850-8-1 0:0:0.0, 1850-9-1 0:0:0.0,
1850-10-1 0:0:0.0, 1850-11-1 0:0:0.0, 1850-12-1 0:0:0.0, 1851-1-1 0:0:0.0]
```

**Comment by doutriaux1** [comment]
```python
import cdms2

import os
import cdms2

path = '/p/cscratch/acme/data/test_model_data_for_acme_diags/time-series/E3SM_v1'
# For NERSC, use the path below:
# path = '/global/project/projectdirs/acme/acme_diags/test_model_data_for_acme_diags/time-series/E3SM_v1/'
file_name = 'PRECC_185001_201312.nc'
var = 'PRECC'

with cdms2.open(os.path.join(path, file_name)) as f:
    start, end = 1850, 1850
    s_time = '{}-1-15'.format(start)
    e_time = '{}-12-15'.format(end)
    v = f(var, time=(s_time, e_time, "ccb"))
    times = v.getTime().asComponentTime()
    print('The num of times were {} when it should be {}.'.format(len(times), (end-start+1)*12))
    print(times)
```

**Comment by doutriaux1** [comment]
The following works


import os
import cdms2

path = '/p/cscratch/acme/data/test_model_data_for_acme_diags/time-series/E3SM_v1'
# For NERSC, use the path below:
# path = '/global/project/projectdirs/acme/acme_diags/test_model_data_for_acme_diags/time-series/E3SM_v1/'
file_name = 'PRECC_185001_201312.nc'
var = 'PRECC'

with cdms2.open(os.path.join(path, file_name)) as f:
    start, end = 1850, 1850
    s_time = '{}-1-15'.format(start)
    e_time = '{}-12-15'.format(end)
    v = f(var, time=(s_time, e_time, "ccb"))
    times = v.getTime().asComponentTime()
    print('The num of times were {} when it should be {}.'.format(len(times), (end-start+1)*12))
    print(times)



On 5/23/19 3:23 PM, Zeshawn Shaheen wrote:

In the above code, when we use v = f(var, time=slice(0, 12)), I'm still not getting January of 1850, but am getting January 1851, since I'm getting 12 slices. See below:

[1850-2-1 0:0:0.0, 1850-3-1 0:0:0.0, 1850-4-1 0:0:0.0, 1850-5-1 0:0:0.0,
1850-6-1 0:0:0.0, 1850-7-1 0:0:0.0, 1850-8-1 0:0:0.0, 1850-9-1 0:0:0.0,
1850-10-1 0:0:0.0, 1850-11-1 0:0:0.0, 1850-12-1 0:0:0.0, 1851-1-1 0:0:0.0]

**Comment by zshaheen** [comment]
I can also confirm that using `'ccb'` when slicing the data works. This was fixed in 8557a6090375230b5bdf3f71cc2301a10339f923. Thank you Charles and cdms.

**Comment by doutriaux1** [comment]
[mysterious_3rd_arg.pdf](https://github.com/E3SM-Project/e3sm_diags/files/3214437/mysterious_3rd_arg.pdf)
